pkg/vault/contracts/AddUnbalancedLiquidityViaSwapRouter.sol

```diff
@@ -0,0 +1,228 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IPermit2 } from "permit2/src/interfaces/IPermit2.sol";
+
+import { IWETH } from "@balancer-labs/v3-interfaces/contracts/solidity-utils/misc/IWETH.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import {
+    IAddUnbalancedLiquidityViaSwapRouter
+} from "@balancer-labs/v3-interfaces/contracts/vault/IAddUnbalancedLiquidityViaSwapRouter.sol";
+import "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+import "@balancer-labs/v3-interfaces/contracts/vault/RouterTypes.sol";
+
+import { RouterHooks } from "./RouterHooks.sol";
+
+/**
+ * @notice Enable adding and removing liquidity unbalanced on pools that do not support it natively.
+ * @dev This extends the standard `Router` in order to call shared internal hook implementation functions.
+ * It factors out the unbalanced adds into two operations: a proportional add and a swap, executes them using
+ * the standard router, then checks the limits.
+ */
+contract AddUnbalancedLiquidityViaSwapRouter is RouterHooks, IAddUnbalancedLiquidityViaSwapRouter {
```

Comment:
Late comment: in the tests we're using the regular router to query swaps and adds individually.

Maybe it's worth inheriting from `RouterQueries` here instead, making those entrypoints available?